### DISCUSSION TITLE
Release a standard hpp/cpp library

### DISCUSSION THREAD

**Discussion by ZigRazor** [ZigRazor/CXXGraph#362] [Ideas]
Since the project is scaling up with dimensions, and the feature are ever more.
As suggested by @nrkramer  maybe think a release of standard hpp/cpp library could be a good idea.
The Drawbacks of header-only library are starting to manifest in this library.

We can think a restricted header-only library for basic function for the common use, and a hpp/cpp for full funtionalities?

What do you think @nrkramer @sbaldu ?

**Comment by ZigRazor** [comment]
How do you plan to maintain generality?

**Reply by ZigRazor** [reply]
Also Boost Graph library is "header-only" to maintain the generatlity as the STL.

**Comment by nrkramer** [comment]
One drawback I'm already seeing is a big increase in compilation time. Unfortunately there's no workaround for that until C++20, which introduces modules.

**Reply by nrkramer** [reply]
If we do decide to pursue C++20, the ability to backport bug fixes and features becomes necessary - which is a lot of work.

It might be easier to make sure that we're also C++14,17 compatible - or have some policy of backwards compatibility that is reasonable.

**Reply by ZigRazor** [reply]
I Think some `#ifdef` can help us to go towards the `C++20` standard.
Obviously we need to maintain the retrocompatibility with the old standards.
This maybe have a cost of work, but if the `#ifdef` macro are well placed I think the effort can be afforded.

**Comment by nrkramer** [comment]
Header-only does have a few of advantages:
- **Built-in template metaprogramming support:** Don't need a bunch of cross-compilation unit template foisting.
- **Ease of distribution:** Just include the header.
- **Better opportunity for function inlining:** Because everything is in the same compilation units, the compiler can easily inline. Although this is largely replicable with LTO (link-time optimization)
- **Ease of development:** Don't need to worry about library compilation, optimization flags, etc
